src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java

```diff
@@ -2236,60 +2237,53 @@ public Type visitErrorType(ErrorType t, Symbol sym) {
         };
 
     /**
-     * Return the base type of t or any of its outer types that starts
-     * with the given symbol.  If none exists, return null.
+     * Return the base type of t or any of its outer types (by following the types enclosing type)
+     * that start with the given symbol.
+     * If none exists, return null.
      *
      * @param t a type
      * @param sym a symbol
      */
     public Type asOuterSuper(Type t, Symbol sym) {
-        switch (t.getTag()) {
-        case CLASS:
-            do {
-                Type s = asSuper(t, sym);
-                if (s != null) return s;
-                t = t.getEnclosingType();
-            } while (t.hasTag(CLASS));
-            return null;
-        case ARRAY:
-            return isSubtype(t, sym.type) ? sym.type : null;
-        case TYPEVAR:
-            return asSuper(t, sym);
-        case ERROR:
-            return t;
-        default:
-            return null;
-        }
+        return asOuter(t, sym, Type::getEnclosingType);
     }
 
     /**
-     * Return the base type of t or any of its enclosing types that
-     * starts with the given symbol.  If none exists, return null.
+     * Return the base type of t or any of its outer types (by following owner's
+     * enclosing class) that start with the given symbol.
+     * If none exists, return null.
      *
      * @param t a type
      * @param sym a symbol
      */
     public Type asEnclosingSuper(Type t, Symbol sym) {
-        switch (t.getTag()) {
-        case CLASS:
-            do {
-                Type s = asSuper(t, sym);
-                if (s != null) return s;
-                Type outer = t.getEnclosingType();
-                t = (outer.hasTag(CLASS)) ? outer :
-                    (t.tsym.owner.enclClass() != null) ? t.tsym.owner.enclClass().type :
-                    Type.noType;
-            } while (t.hasTag(CLASS));
-            return null;
-        case ARRAY:
-            return isSubtype(t, sym.type) ? sym.type : null;
-        case TYPEVAR:
-            return asSuper(t, sym);
-        case ERROR:
-            return t;
-        default:
-            return null;
+        return asOuter(t, sym, type -> getOwnerEnclosingClassType(type));
+    }
+
+    private static Type getOwnerEnclosingClassType(Type type) {
+        return (type.tsym.owner.enclClass() != null)
+                ? type.tsym.owner.enclClass().type
+                : Type.noType;
+    }
+
+    /**
+     * Return the (most specific) base type of t that starts with
```

Comment:
Base type smells like "raw" type -- e.g. non-parameterized. I wonder if it would be clearer to specify in a more direct way:
```
Calls `asSuper(S, sym)` on a sequence of types until a match is found. The sequence of types starts with `t` and the next type in the sequence is obtained by passing the previous type in the sequence to the unary operator `nextType`.
```

Reply:
The text I wrote above can also be "specialized" for `asOuterSuper`:

```
Calls `asSuper(S, sym)` on a sequence of types until a match is found. The sequence of types starts with `t` and the next type in the sequence is obtained by calling `getEnclosingType()` on  the previous type in the sequence.
```

And similarly, for `asEnclosingSuper`:

```
Calls `asSuper(S, sym)` on a sequence of types until a match is found. The sequence of types starts with `t` and the next type in the sequence is obtained by obtaining innermost lexically enclosing class type of the previous type in the sequence.
```